### ISSUE TITLE
Add support for breakpoint dependencies (trigger points)

### ISSUE THREAD

**Issue by gayanper** [eclipse-theia/theia#11969]
When debugging a production level application like an angular app, java microservice we often need to add breakpoints into methods which are called frequently. For an example rendering method in the frontend or event subscriber on the consumer. But most of the time we are only interested to suspend on these methods when it was triggered from a another operation which we are interested of.

How this is done in Eclipse is using trigger point, We can add a trigger point (different kind of a breakpoint) on the operation which is called before our debugging method, so once only this trigger point is hit, our rendering or event subscription breakpoint will get enabled. In IntelliJ this is done using breakpoint dependencies.

having this support for vscode would be really beneficial, if we can build this on the vscode debug extension layer, we could even support polygot scenarios, where the trigger point is in one app we are debugging and actual breakpoint which get enabled might be on another app.

**Comment by gayanper** [comment]
For more information such as what is implemented so far can be found in https://github.com/microsoft/vscode/issues/152428 with screencasts and images

**Comment by JonasHelming** [comment]
@gayanper If I understood correctly from your Twitter message, you would be interested in contributing this to Theia, if there is interest from the community? @marcdumais-work

**Comment by gayanper** [comment]
> @gayanper If I understood correctly from your Twitter message, you would be interested in contributing this to Theia, if there is interest from the community? @marcdumais-work

Correct 🙂

**Comment by gayanper** [comment]
> Yes, I think this sounds interesting.
> 
> Gayan, if I understand correctly, using a _trigger point_ could be less performance-affecting in busy code, versus using a conditional breakpoint, that would need to be re-evaluated each time execution passes-by the (event subscription) breakpoint?

Yes it will be more efficient since there is no need to perform evaluations on the application runtime using the actual language debug protocol to check if the debut point is applicable, The trigger points are based on events that are received by the DAP client which is vscode or theia-ide.

> 
> Additionally, it seems it could often be much easier to use a _trigger point_, to very precisely control the scenario in which a breakpoint is active, versus using a conditional breakpoint (e.g. based on local variable or function call parameter value).
> 
> > we could even support polygot scenarios, where the trigger point is in one app we are debugging and actual breakpoint which get enabled might be on another app
> 
> In some debug scenarios, a host and a target system might be debugged in parallel. e.g. debugging a program running on GPU and its host counter-part running on the CPU. Each might have its own Debug Adapter and debug session, running in the same IDE. Would that be a "polyglot scenario"?

Yes you could consider it as well, But also it could be a client and server which are written in two different languages as well.

**Comment by gayanper** [comment]
Hi Devs, what do you think about this feature proposal ?

**Comment by JonasHelming** [comment]
@gayanper Most devs are probably on vacation, so it might take a bit until @marcdumais-work get back to your comments
